source/s3_meta_request.c

```diff
@@ -0,0 +1,673 @@
+    if (meta_request_result->error_response_headers) {
+        header_list = PyList_New(aws_http_headers_count(meta_request_result->error_response_headers));
+        if (!header_list) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+            goto done;
+        }
+        if (s_get_py_headers(meta_request_result->error_response_headers, header_list)) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+            goto done;
+        }
+    }
+    if (meta_request_result->error_response_body) {
+        error_body = *(meta_request_result->error_response_body);
+    }
+    result = PyObject_CallMethod(
+        request_binding->self_py,
+        "_on_finish",
+        "(iOy#)",
+        meta_request_result->error_code,
+        header_list ? header_list : Py_None,
+        (const char *)(error_body.buffer),
+        (Py_ssize_t)error_body.len);
+
+    if (result) {
+        Py_DECREF(result);
+    } else {
+        /* Callback might fail during application shutdown */
+        PyErr_WriteUnraisable(request_binding->self_py);
+    }
+done:
+    Py_XDECREF(header_list);
+    Py_CLEAR(request_binding->self_py);
+    Py_CLEAR(request_binding->http_message_py);
+    PyGILState_Release(state);
+    /*************** GIL RELEASE ***************/
+}
+
+/* Invoked when the python object get cleaned up */
+static void s_s3_meta_request_capsule_destructor(PyObject *capsule) {
+    struct s3_meta_request_binding *meta_request = PyCapsule_GetPointer(capsule, s_capsule_name_s3_meta_request);
+    AWS_FATAL_ASSERT(!meta_request->release_called);
+    Py_XDECREF(meta_request->self_py);
+    Py_XDECREF(meta_request->http_message_py);
+
+    if (meta_request->recv_file) {
+        fclose(meta_request->recv_file);
+    }
+
+    aws_s3_meta_request_release(meta_request->native);
+
+    meta_request->release_called = true;
+
+    s_destroy_if_ready(meta_request);
+}
+
+/* Callback from C land, invoked when the underlying shutdown process finished */
+static void s_s3_request_on_shutdown(void *user_data) {
+    struct s3_meta_request_binding *request_binding = user_data;
+
+    /*************** GIL ACQUIRE ***************/
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
+        return; /* Python has shut down. Nothing matters anymore, but don't crash */
+    }
+
+    request_binding->shutdown_called = true;
+
+    bool destroy_after_shutdown = request_binding->release_called;
+
+    /* Invoke on_shutdown, then clear our reference to it */
+    PyObject *result = PyObject_CallFunction(request_binding->on_shutdown, NULL);
+    if (result) {
+        Py_DECREF(result);
+    } else {
+        /* Callback might fail during application shutdown */
+        PyErr_WriteUnraisable(PyErr_Occurred());
+    }
+    Py_CLEAR(request_binding->on_shutdown);
+    Py_CLEAR(request_binding->py_core);
+
+    if (destroy_after_shutdown) {
+        aws_mem_release(aws_py_get_allocator(), request_binding);
+    }
```

Comment:
shouldn't we just be calling `s_destroy_if_ready()`?
having cleanup code in many different places is ... error prone as time goes on and more people touch this code